File: use-cases/solicitations/code/csharp/src/sp-api-csharp-app/lambda/SubscribeNotificationsHandler.cs
using Amazon.Lambda.Core;
using Amazon.Scheduler;
using Amazon.Scheduler.Model;
using Amazon.SellingPartnerAPIAA;
using Amazon.Lambda;
using SpApiCsharpApp;
using Newtonsoft.Json;
using SpApiCsharpApp.swaggerClient.Api;
using SpApiCsharpApp.swaggerClient.Model;
using System;
using Amazon.Runtime.Internal.Util;
using SpApiCsharpApp.swaggerClient.Model.Solicitations;
using Amazon.SQS;
using SpApiCsharpApp.swaggerClient.Model.Notifications;
using System.ComponentModel;
using SpApiCsharpApp.swaggerClient.Client;
using SpApiCsharpApp.lambda.utils;

// Assembly attribute to enable the Lambda function's JSON input to be converted into a .NET class.
//[assembly: LambdaSerializer(typeof(Amazon.Lambda.Serialization.SystemTextJson.DefaultLambdaJsonSerializer))]
namespace SpApiCsharpApp
{
    public class SubscribeNotificationsHandler
    {
        private static String NotificationTypeKeyName = "NotificationType";
        private static String RegionCodeKeyName = "RegionCode";
        private static String ScheduleKeyName = "ScheduleName";

        /* Sample event input
        {
            "NotificationType": "ORDER_CHANGE",
            "RegionCode": "NA",
            "RefreshToken": "Atzr|IwEBIFdeNQT9QH3..."
        }
        */
        public string FunctionHandler(SubscribeNotificationsInput input, ILambdaContext context)
        {
            ILambdaLogger logger = context.Logger;
            logger.LogLine("Subscribe Notifications handler started");
            logger.LogLine(JsonConvert.SerializeObject(input));

            // Retrieve request details from input payload
            var notificationType = input.NotificationType;
            var regionCode = input.RegionCode;
            var refreshToken =input.RefreshToken;

            // Create destination if it doesn't exist
            string destinationId;
            try
            {
                destinationId = CreateDestination(regionCode, refreshToken);
                logger.LogLine($"Destination created - Destination Id: {destinationId}");
            }
            catch (Exception e)
            {
                throw new Exception("Create destination failed", e);
            }

            // Create subscription if it doesnt exist
            string subscriptionId;
            try
            {
                subscriptionId = CreateSubscription(regionCode, refreshToken, notificationType, destinationId);
                logger.LogLine($"Subscription created - Subscription Id: {subscriptionId}");

                return $"Destination Id: {destinationId} - Subscription Id: {subscriptionId}";
            }
            catch (Exception e)
            {
                throw new Exception("Create subscription failed", e);
            }

        }

        private String? CreateDestination(string regionCode, string refreshToken)
        {
            // Get SQS queue ARN from environment variable
             var sqsQueueArn = Environment.GetEnvironmentVariable(Constants.SqsQueueArnEnvVariable);

            // Create request parameters for "CreateDestination" API call
            var sqsResource = new SqsResource (sqsQueueArn);
            var resourceSpec = new DestinationResourceSpecification (sqsResource);

            var request = new CreateDestinationRequest(
            
                name:Guid.NewGuid().ToString(),
                resourceSpecification: resourceSpec
            );

            //Invoke the Notifications API using a grantless notifications scope
            NotificationsApi notificationsApi = ApiUtils.GetNotificationsApi(regionCode, refreshToken, Constants.LwaNotificationsScope);

            var destinationId = "";
            try
            {
                CreateDestinationResponse response = notificationsApi.CreateDestination(request); 
                destinationId = response.Payload.DestinationId;
            }
            catch (ApiException e)
            {
                // If destination exists, retrieve ID
                if (e!=null && e.ErrorCode==409)
                {
                    var destinations = notificationsApi.GetDestinations();

                    var sqsDestination = destinations!=null?destinations.Payload.FirstOrDefault(d => d.Resource.Sqs != null
                            && sqsQueueArn == d.Resource.Sqs.Arn):null;

                    if (sqsDestination != null)
                    {
                        destinationId = sqsDestination.DestinationId;
                    }
                }
                else
                {
                    throw;
                }
            }

            return destinationId ;
        }

        private String? CreateSubscription(string regionCode, string refreshToken, string notificationType, string destinationId)
        {
            // Create request parameters for "CreateSubscription" API call
            var request = new CreateSubscriptionRequest(
                destinationId: destinationId,
                payloadVersion: "1.0",
                processingDirective: new ProcessingDirective(new EventFilter(eventFilterType: EventFilter.EventFilterTypeEnum.ORDERCHANGE))
            ) ;

            // Call Notifications API using refresh token
            NotificationsApi notificationsApi = ApiUtils.GetNotificationsApi(regionCode, refreshToken);

            try
            {
                CreateSubscriptionResponse response = notificationsApi.CreateSubscription(request, notificationType);
                return response.Payload.SubscriptionId;
            }
            catch (ApiException e)
            {
                // If subscription exists, retrieve ID
                if (e != null && e.ErrorCode == 409)                   
                {
                    var response = notificationsApi.GetSubscription(notificationType);
                    return response.Payload.SubscriptionId;
                }
                else
                {
                    throw;
                }
            }
        }

    }

}

File: use-cases/solicitations/code/python/src/subscribe_notifications_handler.py
from api_utils import ApiUtils
from constants import Constants

import os
import uuid
import logging

from api_models.notifications_api.swagger_client.rest import ApiException
from api_models.notifications_api.swagger_client.models.create_destination_request import CreateDestinationRequest
from api_models.notifications_api.swagger_client.models.destination_resource_specification import DestinationResourceSpecification
from api_models.notifications_api.swagger_client.models.sqs_resource import SqsResource
from api_models.notifications_api.swagger_client.models.create_subscription_request import CreateSubscriptionRequest

'''
# Sample event input
{
    "NotificationType": "ORDER_CHANGE",
    "RegionCode": "NA",
    "RefreshToken": "Atzr|IwEBIFdeNQT9QH3..."
}
'''

def lambda_handler(event, context):
    logger = logging.getLogger()
    logger.setLevel(logging.INFO)
    logger.info('Subscribe Notifications handler started')
    logger.info(event)

    # Retrieve request details from the input payload
    notification_type = event.get('NotificationType')
    region_code = event.get('RegionCode')
    refresh_token = event.get('RefreshToken')

    # Create destination if it doesn't exist
    try:
        destination_id = create_destination(region_code, refresh_token)
        logger.info(f'Destination created - Destination Id: {destination_id}')
    except Exception as e:
        raise Exception('Create destination failed', e)

    # Create subscription
    try:
        subscription_id = create_subscription(region_code, refresh_token, notification_type, destination_id)
        logger.info(f'Subscription created - Subscription Id: {subscription_id}')

        return f'Destination Id: {destination_id} - Subscription Id: {subscription_id}'
    except Exception as e:
        raise Exception('Create subscription failed', e)

def create_destination(region_code, refresh_token):
    # Get the SQS arn from the Lambda function's environment variables
    sqs_queue_arn = os.environ[Constants.SQS_QUEUE_ARN_ENV_VARIABLE]

    sqs_resource = SqsResource(arn=sqs_queue_arn)
    resource_spec = DestinationResourceSpecification(sqs=sqs_resource)
    request = CreateDestinationRequest(name=str(uuid.uuid4()), resource_specification=resource_spec)

    # Invoke the Notifications API using a grantless notifications scope
    api_utils = ApiUtils(refresh_token, region_code, 'notifications', 'No', Constants.LWA_NOTIFICATIONS_SCOPE)

    destination_id = ''
    try:
        create_destination_response = api_utils.call_notifications_api(
            method='create_destination',
            body=request)
        destination_id = create_destination_response.payload.destination_id
    except Exception as e:
        # If the destination already exists, retrieve the destination Id
        if e.status == 409:
            get_destinations_response = api_utils.call_notifications_api(method='get_destinations')

            sqs_destination = next(
                (destination for destination in get_destinations_response.payload if
                 destination['resource']['sqs'] and sqs_queue_arn == destination['resource']['sqs']['arn']), None)

            if sqs_destination:
                destination_id = sqs_destination['destinationId']
        else:
            raise e

    return destination_id


def create_subscription(region_code, refresh_token, notification_type, destination_id):
    request = CreateSubscriptionRequest(payload_version='1.0', destination_id=destination_id)

    api_utils = ApiUtils(refresh_token, region_code, 'notifications')

    try:
        response = api_utils.call_notifications_api(
            method='create_subscription',
            body=request,
            notification_type=notification_type)
        subscription_id = response.payload.subscription_id
    except ApiException as e:
        # If a subscription for the notification types already exists, retrieve the subscription Id
        if e.status == 409:
            get_subscription_response = api_utils.call_notifications_api(
                method='get_subscription',
                notification_type=notification_type)
            subscription_id = get_subscription_response.payload.subscription_id
        else:
            raise e

    return subscription_id
